Support token with scopes

Hello folks,

I noticed that using Confluence Cloud, plugin version 0.30.1, only tokens **without** scopes are supported by the plugin.

When using a token with scopes (even if all scopes are enabled), an error 401 is raised.

I suspect this is because URLs of the APIs are different when using a token with scopes as per [this documentation](https://support.atlassian.com/atlassian-account/docs/manage-api-tokens-for-your-atlassian-account/):

> ```
> ## URL for API Tokens
> curl -v https://mysite.atlassian.net --user me@example.com:my-api-token
> 
> ## URLs for API tokens with scopes for Confluence
> curl -v https://api.atlassian.com/ex/confluence/{cloudId} --user me@example.com:my-api-token
> ```
> In this example, me@example.com is the email address for the Atlassian account you're using to create the token.

---

Partially related to:
- https://github.com/confluence-publisher/confluence-publisher/issues/487
- https://github.com/confluence-publisher/confluence-publisher/issues/484#issuecomment-3704924130

@cstettler I haven't had time to dig yet but I'm still getting the following error:

```
Caused by: org.sahli.asciidoc.confluence.publisher.client.http.RequestFailedException: request failed (request: GET https://mycompany.atlassian.net/wiki/api/v2/pages/11381940?include-version=true <empty body>, response: 401 Unauthorized <!doctype html><html lang="en"><head><title>HTTP Status 401 – Unauthorized</title><style type="text/css">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 401 – Unauthorized</h1></body></html>, reason: <none>)
    at org.sahli.asciidoc.confluence.publisher.client.http.ConfluenceRestV2Client.lambda$sendRequestAndFailIfNot20x$9 (ConfluenceRestV2Client.java:213)
    at org.sahli.asciidoc.confluence.publisher.client.http.ConfluenceRestV2Client.sendRequest (ConfluenceRestV2Client.java:228)
    at org.sahli.asciidoc.confluence.publisher.client.http.ConfluenceRestV2Client.sendRequestAndFailIfNot20x (ConfluenceRestV2Client.java:210)
    at org.sahli.asciidoc.confluence.publisher.client.http.ConfluenceRestV2Client.getPageWithContentAndVersionById (ConfluenceRestV2Client.java:190)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.updatePage (ConfluencePublisher.java:190)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.startPublishingReplacingAncestorId (ConfluencePublisher.java:114)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.publish (ConfluencePublisher.java:85)
```

It seems to still be using the URL for non-scoped tokens.

Maybe I'm not using the right SNAPSHOT though, they're all published as `0.0.0-SNAPSHOT` so there's a small risk I'm not using the good one. From the logs it's `0.0.0-20260517.172750-20`.

Ah thanks, I missed that @cstettler :)

Seems better but still some errors.

(for the record) With a token I generated previously giving it all "Classic" scopes (there are "classic" and "granular" scopes):
```
Caused by: org.sahli.asciidoc.confluence.publisher.client.http.RequestFailedException: Unauthorized; scope does not match (request: GET https://api.atlassian.com/ex/confluence/xxx/api/v2/pages/11381940?include-version=true <empty body>, response: 401 Unauthorized {"code":401,"message":"Unauthorized; scope does not match"}, reason: <none>)
    at org.sahli.asciidoc.confluence.publisher.client.http.ConfluenceRestV2Client.lambda$sendRequestAndFailIfNot20x$9 (ConfluenceRestV2Client.java:213)
```

And with the scopes listed in your documentation:
```
Caused by: org.sahli.asciidoc.confluence.publisher.client.http.RequestFailedException: No message available (request: GET https://api.atlassian.com/wiki/api/v2/pages/11382313/attachments?limit=2&cursor=xxx== <empty body>, response: 404 Not Found {"timestamp":"2026-05-19T18:10:14.698736667Z","status":404,"error":"Not Found","message":"No message available","path":"/wiki/api/v2/pages/11382313/attachments"}, reason: <none>)
```

Not sure what I could give you as additional details for you to troubleshoot this 🤷

Complete stack:
```
Caused by: org.sahli.asciidoc.confluence.publisher.client.http.RequestFailedException: No message available (request: GET https://api.atlassian.com/wiki/api/v2/pages/11382313/attachments?limit=2&cursor=xxx== <empty body>, response: 404 Not Found {"timestamp":"2026-05-19T18:10:14.698736667Z","status":404,"error":"Not Found","message":"No message available","path":"/wiki/api/v2/pages/11382313/attachments"}, reason: <none>)
    at org.sahli.asciidoc.confluence.publisher.client.http.ConfluenceRestV2Client.lambda$sendRequestAndFailIfNot20x$9 (ConfluenceRestV2Client.java:213)
    at org.sahli.asciidoc.confluence.publisher.client.http.ConfluenceRestV2Client.sendRequest (ConfluenceRestV2Client.java:228)
    at org.sahli.asciidoc.confluence.publisher.client.http.ConfluenceRestV2Client.sendRequestAndFailIfNot20x (ConfluenceRestV2Client.java:210)
    at org.sahli.asciidoc.confluence.publisher.client.http.ConfluenceRestV2Client.getAttachments (ConfluenceRestV2Client.java:260)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.deleteConfluenceAttachmentsNotPresentUnderPage (ConfluencePublisher.java:161)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.lambda$startPublishingUnderAncestorId$1 (ConfluencePublisher.java:135)
    at java.util.ArrayList.forEach (ArrayList.java:1511)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.startPublishingUnderAncestorId (ConfluencePublisher.java:129)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.lambda$startPublishingUnderAncestorId$1 (ConfluencePublisher.java:138)
    at java.util.ArrayList.forEach (ArrayList.java:1511)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.startPublishingUnderAncestorId (ConfluencePublisher.java:129)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.startPublishingReplacingAncestorId (ConfluencePublisher.java:121)
    at org.sahli.asciidoc.confluence.publisher.client.ConfluencePublisher.publish (ConfluencePublisher.java:85)
    at org.sahli.asciidoc.confluence.publisher.maven.plugin.AsciidocConfluencePublisherMojo.execute (AsciidocConfluencePublisherMojo.java:184)
```

I confirm the rootUrl does contain the cloudId, you can see it in the 1st error I got but it's indeed missing from this stacktrace.

--- a/asciidoc-confluence-publisher-client/src/it/java/org/sahli/asciidoc/confluence/publisher/client/ConfluencePublisherV1IntegrationTest.java
+++ b/asciidoc-confluence-publisher-client/src/it/java/org/sahli/asciidoc/confluence/publisher/client/ConfluencePublisherV1IntegrationTest.java
@@ -141,6 +141,35 @@ public class ConfluencePublisherV1IntegrationTest {
     }
 
     @Test
+    public void publish_pageWithMoreThanPageSizeAttachmentsAndAttachmentsRemovedOnRepublish_removesAttachmentsFromPage() {
+        // arrange
+        String title = uniqueTitle("Single Page");
+        Map<String, String> attachments = new HashMap<>();
+        attachments.put("attachmentOne.txt", absolutePathTo("attachments/attachmentOne.txt"));
+        attachments.put("attachmentTwo.txt", absolutePathTo("attachments/attachmentTwo.txt"));
+        attachments.put("attachmentThree.txt", absolutePathTo("attachments/attachmentThree.txt"));
+
+        ConfluencePageMetadata confluencePageMetadata = confluencePageMetadata(title, absolutePathTo("single-page/single-page.xhtml"), attachments);
+        ConfluencePublisherMetadata confluencePublisherMetadata = confluencePublisherMetadata(confluencePageMetadata);
+        ConfluencePublisher confluencePublisher = confluencePublisher(confluencePublisherMetadata, REPLACE_ANCESTOR);
+
+        confluencePublisher.publish();
+
+        confluencePageMetadata = confluencePageMetadata(title, absolutePathTo("single-page/single-page.xhtml"));
+        confluencePublisherMetadata = confluencePublisherMetadata(confluencePageMetadata);
+        confluencePublisher = confluencePublisher(confluencePublisherMetadata, REPLACE_ANCESTOR);
+
+        // act
+        confluencePublisher.publish();
+
+        // assert
+        givenAuthenticatedAsPublisher()
+            .when().get(rootPageAttachments())
+            .then()
+            .body("results", hasSize(0));
+    }
+
+    @Test
     public void publish_attachmentWithNonAsciiCharactersInTitle_uploadsAttachmentWithCorrectName() {
         // arrange
         String title = uniqueTitle("Single Page");

--- a/asciidoc-confluence-publisher-client/src/it/java/org/sahli/asciidoc/confluence/publisher/client/ConfluencePublisherV2IntegrationTest.java
+++ b/asciidoc-confluence-publisher-client/src/it/java/org/sahli/asciidoc/confluence/publisher/client/ConfluencePublisherV2IntegrationTest.java
@@ -141,6 +141,35 @@ public class ConfluencePublisherV2IntegrationTest {
     }
 
     @Test
+    public void publish_pageWithMoreThanPageSizeAttachmentsAndAttachmentsRemovedOnRepublish_removesAttachmentsFromPage() {
+        // arrange
+        String title = uniqueTitle("Single Page");
+        Map<String, String> attachments = new HashMap<>();
+        attachments.put("attachmentOne.txt", absolutePathTo("attachments/attachmentOne.txt"));
+        attachments.put("attachmentTwo.txt", absolutePathTo("attachments/attachmentTwo.txt"));
+        attachments.put("attachmentThree.txt", absolutePathTo("attachments/attachmentThree.txt"));
+
+        ConfluencePageMetadata confluencePageMetadata = confluencePageMetadata(title, absolutePathTo("single-page/single-page.xhtml"), attachments);
+        ConfluencePublisherMetadata confluencePublisherMetadata = confluencePublisherMetadata(confluencePageMetadata);
+        ConfluencePublisher confluencePublisher = confluencePublisher(confluencePublisherMetadata, REPLACE_ANCESTOR);
+
+        confluencePublisher.publish();
+
+        confluencePageMetadata = confluencePageMetadata(title, absolutePathTo("single-page/single-page.xhtml"));
+        confluencePublisherMetadata = confluencePublisherMetadata(confluencePageMetadata);
+        confluencePublisher = confluencePublisher(confluencePublisherMetadata, REPLACE_ANCESTOR);
+
+        // act
+        confluencePublisher.publish();
+
+        // assert
+        givenAuthenticatedAsPublisher()
+            .when().get(rootPageAttachments())
+            .then()
+            .body("results", hasSize(0));
+    }
+
+    @Test
     public void publish_attachmentWithNonAsciiCharactersInTitle_uploadsAttachmentWithCorrectName() {
         // arrange
         String title = uniqueTitle("Single Page");

--- a/asciidoc-confluence-publisher-client/src/it/resources/attachments/attachmentThree.txt
+++ b/asciidoc-confluence-publisher-client/src/it/resources/attachments/attachmentThree.txt
@@ -1,0 +1,1 @@
+attachment3

--- a/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestFactory.java
+++ b/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestFactory.java
@@ -48,8 +48,6 @@ public interface HttpRequestFactory {
 
     HttpGet getAttachmentsRequest(String contentId, Integer limit, Integer start, String expandOptions);
 
-    HttpGet getAttachmentContentRequest(String relativeDownloadLink);
-
     HttpGet getPropertyByKeyRequest(String contentId, String key);
 
     HttpDelete deletePropertyByKeyRequest(String contentId, String key);

--- a/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV1Factory.java
+++ b/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV1Factory.java
@@ -256,14 +256,7 @@ class HttpRequestV1Factory implements HttpRequestFactory {
         return getAttachmentsRequest;
     }
 
-    @Override
-    public HttpGet getAttachmentContentRequest(String relativeDownloadLink) {
-        assertMandatoryParameter(isNotBlank(relativeDownloadLink), "relativeDownloadLink");
-
-        return new HttpGet(this.rootConfluenceUrl + relativeDownloadLink);
-    }
-
-    @Override
+  @Override
     public HttpGet getPropertyByKeyRequest(String contentId, String key) {
         assertMandatoryParameter(isNotBlank(contentId), "contentId");
         assertMandatoryParameter(isNotBlank(key), "key");

--- a/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV2Factory.java
+++ b/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV2Factory.java
@@ -257,7 +257,14 @@ class HttpRequestV2Factory implements HttpRequestFactory {
     public HttpGet getNextAttachmentsRequest(String nextLink) {
         assertMandatoryParameter(isNotBlank(nextLink), "nextLink");
 
-        return new HttpGet(this.confluenceServerUrl + nextLink);
+        String nextLinkBaseUrl = this.rootConfluenceUrl;
+
+        // next link contains /wiki path segment already present in root confluence url (only for api tokens, not for scoped tokens)
+        if (nextLinkBaseUrl.endsWith("/wiki")) {
+            nextLinkBaseUrl = nextLinkBaseUrl.substring(0, nextLinkBaseUrl.length() - "/wiki".length());
+        }
+
+        return new HttpGet(nextLinkBaseUrl + nextLink);
     }
 
     @Override

--- a/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV2Factory.java
+++ b/asciidoc-confluence-publisher-client/src/main/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV2Factory.java
@@ -268,13 +268,6 @@ class HttpRequestV2Factory implements HttpRequestFactory {
     }
 
     @Override
-    public HttpGet getAttachmentContentRequest(String relativeDownloadLink) {
-        assertMandatoryParameter(isNotBlank(relativeDownloadLink), "relativeDownloadLink");
-
-        return new HttpGet(this.confluenceServerUrl + relativeDownloadLink);
-    }
-
-    @Override
     public HttpGet getPropertyByKeyRequest(String contentId, String key) {
         assertMandatoryParameter(isNotBlank(contentId), "contentId");
         assertMandatoryParameter(isNotBlank(key), "key");

--- a/asciidoc-confluence-publisher-client/src/test/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV1FactoryTest.java
+++ b/asciidoc-confluence-publisher-client/src/test/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV1FactoryTest.java
@@ -583,27 +583,6 @@ public class HttpRequestV1FactoryTest {
     }
 
     @Test
-    public void getAttachmentContentRequest_withValidParameters_returnsHttpGetRequest() {
-        // arrange
-        String relativeDownloadLink = "/download/attachment.txt";
-
-        // act
-        HttpGet getAttachmentContentRequest = this.httpRequestFactory.getAttachmentContentRequest(relativeDownloadLink);
-
-        // assert
-        assertThat(getAttachmentContentRequest.getURI().toString(), is(ROOT_CONFLUENCE_URL + relativeDownloadLink));
-    }
-
-    @Test
-    public void getAttachmentContentRequest_withBlankRelativeDownloadLink_throwsIllegalArgumentException() {
-        // assert
-        assertThrows(IllegalArgumentException.class, () -> {
-            // arrange + act
-            this.httpRequestFactory.getAttachmentContentRequest("");
-        });
-    }
-
-    @Test
     public void setPropertyByKeyRequest_withValidParameters_returnsHttpPostRequest() throws Exception {
         // arrange
         String contentId = "1234";

--- a/asciidoc-confluence-publisher-client/src/test/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV2FactoryTest.java
+++ b/asciidoc-confluence-publisher-client/src/test/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV2FactoryTest.java
@@ -291,27 +291,6 @@ public class HttpRequestV2FactoryTest {
     }
 
     @Test
-    public void getAttachmentContentRequest_withRelativeDownloadLink_returnsUrlWithServerUrlPrefix() {
-        // arrange
-        String relativeDownloadLink = "/download/attachment.txt";
-
-        // act
-        HttpGet request = this.httpRequestFactory.getAttachmentContentRequest(relativeDownloadLink);
-
-        // assert
-        assertThat(request.getURI().toString(), is(ROOT_CONFLUENCE_URL + relativeDownloadLink));
-    }
-
-    @Test
-    public void getAttachmentContentRequest_withBlankRelativeDownloadLink_throwsIllegalArgumentException() {
-        // assert
-        assertThrows(IllegalArgumentException.class, () -> {
-            // arrange + act
-            this.httpRequestFactory.getAttachmentContentRequest("");
-        });
-    }
-
-    @Test
     public void getPropertyByKeyRequest_withValidParameters_returnsHttpGetToApiV2Endpoint() {
         // act
         HttpGet request = this.httpRequestFactory.getPropertyByKeyRequest("1234", "content-hash");
@@ -427,31 +406,5 @@ public class HttpRequestV2FactoryTest {
 
         // assert
         assertThat(request.getURI().toString(), is(ATLASSIAN_SERVER_URL + nextLink));
-    }
-
-    @Test
-    public void getAttachmentContentRequest_withAtlassianApiGatewayUrl_usesServerHostWithoutPathPrefix() {
-        // arrange
-        HttpRequestV2Factory factory = new HttpRequestV2Factory(ATLASSIAN_API_GATEWAY_URL);
-        String relativeDownloadLink = "/ex/confluence/" + ATLASSIAN_API_GATEWAY_CLOUD_ID + "/download/attachments/123/file.png";
-
-        // act
-        HttpGet request = factory.getAttachmentContentRequest(relativeDownloadLink);
-
-        // assert
-        assertThat(request.getURI().toString(), is(ATLASSIAN_SERVER_URL + relativeDownloadLink));
-    }
-
-    @Test
-    public void getAttachmentContentRequest_withStandardConfluenceCloudUrl_usesServerHostWithoutPathPrefix() {
-        // arrange
-        HttpRequestV2Factory factory = new HttpRequestV2Factory("https://mysite.atlassian.net/wiki");
-        String relativeDownloadLink = "/wiki/download/attachments/123/file.png";
-
-        // act
-        HttpGet request = factory.getAttachmentContentRequest(relativeDownloadLink);
-
-        // assert
-        assertThat(request.getURI().toString(), is("https://mysite.atlassian.net" + relativeDownloadLink));
     }
 }

--- a/asciidoc-confluence-publisher-client/src/test/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV2FactoryTest.java
+++ b/asciidoc-confluence-publisher-client/src/test/java/org/sahli/asciidoc/confluence/publisher/client/http/HttpRequestV2FactoryTest.java
@@ -272,7 +272,7 @@ public class HttpRequestV2FactoryTest {
     public void getNextAttachmentsRequest_withRelativeNextLink_prependsConfluenceServerUrl() {
         // arrange
         HttpRequestV2Factory factory = new HttpRequestV2Factory(ROOT_CONFLUENCE_URL);
-        String nextLink = "/api/v2/pages/1234/attachments?cursor=yyy&limit=25";
+        String nextLink = "/wiki/api/v2/pages/1234/attachments?cursor=yyy&limit=25";
 
         // act
         HttpGet request = factory.getNextAttachmentsRequest(nextLink);
@@ -399,12 +399,12 @@ public class HttpRequestV2FactoryTest {
     public void getNextAttachmentsRequest_withAtlassianApiGatewayUrl_prependsServerHostWithoutPath() {
         // arrange
         HttpRequestV2Factory factory = new HttpRequestV2Factory(ATLASSIAN_API_GATEWAY_URL);
-        String nextLink = "/ex/confluence/" + ATLASSIAN_API_GATEWAY_CLOUD_ID + "/api/v2/pages/1234/attachments?cursor=yyy";
+        String nextLink = "/wiki/api/v2/pages/1234/attachments?cursor=yyy";
 
         // act
         HttpGet request = factory.getNextAttachmentsRequest(nextLink);
 
         // assert
-        assertThat(request.getURI().toString(), is(ATLASSIAN_SERVER_URL + nextLink));
+        assertThat(request.getURI().toString(), is(ATLASSIAN_API_GATEWAY_URL + nextLink));
     }
 }

